core/io/json.cpp

```diff
@@ -392,27 +396,24 @@ Error JSON::_parse_array(Array &array, const CharType *p_str, int &index, int p_
 Error JSON::_parse_object(Dictionary &object, const CharType *p_str, int &index, int p_len, int &line, String &r_err_str) {
 
 	bool at_key = true;
-	String key;
+	Variant key;
```

Comment:
Even though the JSON doesn't support arbitrary datatypes as object keys, the key needs to be `Variant` so that previously serialized values (String) can be parsed (see below).